### ISSUE TITLE
Lint/NonLocalExitFromIterator false positive

### ISSUE THREAD

**Issue by bestwebua** [rubocop/rubocop#10386]
Hello! First of all thanks for awesome tool! I think I'm get a false positive from `Lint/NonLocalExitFromIterator` in the code below:

```ruby
def resolve(command_line_args)
  opt_parser = ::OptionParser.new do |parser|
    parser.on('-g', '--upgrade', 'Upgrade to latest version of smtpmock') do
      return if not_installed? # Non-local exit from iterator, without return value. But actually it's a proc-closure

      install_to(current_smtpmock_path[%r{(.+)/.+}, 1])

      self.message = 'smtpmock was upgraded successfully'
    end
    
    self.success = true
  end

  opt_parser.parse(command_line_args)
end
```
--------

## Expected behavior

No flags should be raised.

## Actual behavior

I got `Non-local exit from iterator, without return value` warning. But actually it's a proc-closure.

```
W: Lint/NonLocalExitFromIterator: Non-local exit from iterator, without return value. next, break, Array#find, Array#any?, etc. is preferred.
            return if not_installed?
```

## Steps to reproduce the problem

1. Create a file with the above code snippet as its contents
2. Run Rubocop on it

## RuboCop version

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler.
If you see extension cop versions (e.g. `rubocop-performance`, `rubocop-rspec`, and others)
output by `rubocop -V`, include them as well. Here's an example:

```
$ [bundle exec] rubocop -V
1.25.0 (using Parser 3.1.0.0, rubocop-ast 1.15.1, running on ruby 3.1.0 x86_64-darwin20)
  - rubocop-performance 1.13.2
  - rubocop-rspec 2.8.0
```

**Comment by johnny-miyake** [comment]
I think the name of `Lint/NonLocalExitFromIterator` may be misleading because the cop checks not only iterators but also other methods that take a block. (I think the static check by rubocop can't tell whether a method is an iterator or not.)
https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Lint/NonLocalExitFromIterator
https://github.com/rubocop/rubocop/blob/master/spec/rubocop/cop/lint/non_local_exit_from_iterator_spec.rb
